DOC Changing the dataset from Adult to Diabetes Hospital in Getting Started (#1310)
fairlearn/fairlearn: docs/quickstart.rst

Loading the dataset
^^^^^^^^^^^^^^^^^^^

For this example, we use a `clincial dataset <https://archive.ics.uci.edu/dataset/296/diabetes+130-us+hospitals+for+years+1999-2008>`_ 
of hospital re-admissions over a ten-year period (1998-2008) for 
diabetic patients across 130 different hospitals in the U.S. This scenario 
builds upon prior research on how racial disparities impact health care 
resource allocation in the U.S. For an in-depth analysis of this dataset,
review the `Scipy tutorial <https://github.com/fairlearn/talks/tree/main/2021_scipy_tutorial>`_
that the Fairlearn team presented in 2021.

We will use machine learning to predict whether an individual in the dataset 
is readmnitted to the hospital within 30 days of hospital release. 
A hospital readmission within 30 days can be viewed as a proxy that the 
patients needed more assistance at the release time.
In the next section, we build a classification model to accomplish the 
prediction task.

.. doctest:: quickstart

    >>> import numpy as np
    >>> import pandas as pd
    >>> import matplotlib.pyplot as plt
    >>> data = fetch_diabetes_hospital(as_frame=True)

>     >>> keep_idx = (data.data['gender'] == "Male") | (data.data['gender'] == "Female")

Reviewer: If we're not evaluating by gender we can probably drop this part.

Author: Done!

    :options:  +NORMALIZE_WHITESPACE

    >>> from fairlearn.metrics import MetricFrame
    >>> from sklearn.metrics import accuracy_score, balanced_accuracy_score
    >>> from sklearn.tree import DecisionTreeClassifier
    >>> from sklearn.model_selection import train_test_split
    >>> np.random.seed(42)  # set seed for consistent results
    >>> X_train, X_test, y_train, y_test, A_train, A_test = train_test_split(X_ohe, y, race, random_state=123)
    >>> classifier = DecisionTreeClassifier(min_samples_leaf=10, max_depth=4)
    >>> classifier.fit(X_train, y_train)
    DecisionTreeClassifier(...)
    >>> y_pred = (classifier.predict_proba(X_test)[:,1] >= 0.1)
    >>> mf = MetricFrame(metrics=accuracy_score, y_true=y_test, y_pred=y_pred, sensitive_features=A_test)
    >>> mf.overall
    0.514...
    >>> mf.by_group
    race
    AfricanAmerican    0.530935
    Asian              0.658683
    Caucasian          0.503535
    Hispanic           0.612524
    Other              0.591549
    Unknown            0.574576
    Name: accuracy_score, dtype: float64

> Note that above, because the outcome variable exhibits label imbalance, we set 
> the threshold for triggering a positive prediction lower 
> (to a probability of 0.1 or greater).

Reviewer 2: ```suggestion
Note that our decision threshold for positive predictions is 0.1.
```

    >>> from fairlearn.metrics import MetricFrame
    >>> from sklearn.metrics import accuracy_score
    >>> from sklearn.tree import DecisionTreeClassifier
    >>> from sklearn.model_selection import train_test_split
    >>>
    >>> X_train, X_test, y_train, y_test, \
    A_train, A_test = train_test_split(X_ohe, y, race, random_state=123)
    >>>
    >>> classifier = DecisionTreeClassifier(min_samples_leaf=10, max_depth=4)
    >>> classifier.fit(X_train, y_train)
    DecisionTreeClassifier(...)
    >>> y_pred = (classifier.predict_proba(X_test)[:,1] >= 0.1)
    >>> mf = MetricFrame(metrics=accuracy_score, y_true=y_test, y_pred=y_pred, \
    sensitive_features=A_test)
    >>> mf.overall
    0.529185173538776
    >>> mf.by_group
    race
    AfricanAmerican    0.542303
    Asian              0.690323
    Caucasian          0.520261
    Hispanic           0.595528
    Other              0.561798
    Unknown            0.591525
    Name: accuracy_score, dtype: float64

Note that above, because the outcome variable exhibits label imbalance, we set 
the threshold for triggering a positive prediction lower 

> (to a probability of 0.1 or greater).

Reviewer: I like that you made this adjustment for imbalance but the number 0.1 kind of falls from the sky...

Did you try a few and decide on that or where is it from? 

I checked if DTC has anything for changing the objective, but all I found was the "balanced" class_weight arg https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html
I've never used it but perhaps it could help here?

Adding other @fairlearn/fairlearn-maintainers to chime in if they have ideas.

Author: Yeah, I understand! The proportion of positive labels overall was about 11% so I rounded down, but happy to change it to whatever everyone deems would be a better fit.

Reviewer: Is it possible to just mention that it's set based on the percentage of positive labels in the dataset? That way, it doesn't just occur out of nowhere 🙂

Author: Done! Added a sentence: "We chose this threshold to match the percentage of positive labels in the dataset."

Fairlearn has many standard metrics built-in, such as
selection rate, i.e., the percentage of the population which have '1' as
their label:

.. doctest:: quickstart
    :options:  +NORMALIZE_WHITESPACE

    >>> from fairlearn.metrics import selection_rate

>     >>> sr = MetricFrame(metrics=selection_rate, y_true=y_test, y_pred=y_pred, \

Reviewer: Selection rate is probably not the best criterion here. IMO (and thereby up for discussion) FNR is crucial, or perhaps the entire confusion matrix. FNR captures people that would get readmitted but we don't predict them to be, so that's real (medical) harm... 

FPs are the ones that wouldn't have been readmitted but we kept them in the hospital longer than needed so it's a cost but no harm to the person's health presumably.

So I suggest going with FNR/TPR parity or equalized odds (which captures the whole confusion matrix) as fairness criterion for mitigation with balanced accuracy as objective.

Author: I updated the portion that calculates the selection rate to instead show the false negative rate. I also added a few sentences to provide context to the example. 

I also updated the mitigation section to have the objective of balanced accuracy with equalized odds as the fairness criterion (unless you would prefer to use error rate as the criterion). I do want to quickly note that in the community call we discussed using balanced accuracy as the evaluation criteria in the initial section of the quickstart guide and decided to use accuracy instead with a lower threshold on the prediction probabilities because [to my understanding] balanced accuracy does not differentiate results by subgroups. But let me know if you want me to mock up what unmitigated balanced accuracy would look like, and we can discuss further here.

Reviewer: True. Balanced is balanced by 0/1

    sensitive_features=A_test)
    >>> sr.overall

>     0.5124012420895405

Reviewer: Strongly suggest ellipsis as before because it helps doctest succeed when there are minute differences in the 10th digit etc.

This applies to any number with more than 3 beyond the comma

Author: Done!

>     :start-after: # Analyze metrics using MetricFrame

Author: Yes, I did see that issue. The metrics are imported in plot_quickstart.py, and that should carry through when the file is used in the tutorial. Additionally, in the new tutorial, I don’t believe it uses all the metrics used in the old tutorial. I am not in a place where I can check the documentation in CircleCI now, but I will check it asap and make sure there are no errors.

>     :start-after: # Analyze metrics using MetricFrame

Reviewer 3: cool, happy coding

>     :start-after: # Analyze metrics using MetricFrame

Author: @eroell to address the issue you raised in #1355, the quickstart is not actually pulling plot_quickstart.py from the examples folder - it's pulling from a hidden folder. See below: 
![image](https://github.com/fairlearn/fairlearn/assets/15962391/7faf9fae-0ef9-4d86-a21b-bff10b0cc5cf)
Within the plot_quickstart.py in auto_examples, it is importing all the metrics: 
![image](https://github.com/fairlearn/fairlearn/assets/15962391/6d7d79d8-8e38-4a75-9a3f-176ba792b35f)

In the quickstart in my file, for some reason, the snippet you see is clipped to not show the imports. 

![image](https://github.com/fairlearn/fairlearn/assets/15962391/f2881c5b-fab2-40c2-8b45-21c0ab5dbbc0)

However, your question bring to my attention that the images I changed within auto_examples are not getting updated. I'll work on that today and try to get the images pushed up.

    :start-after: # Analyze metrics using MetricFrame
    :end-before: # Customize plots with ylim

> .. figure:: auto_examples/images/sphx_glr_plot_quickstart_diabetes_001.png

Reviewer 2: I don't see anything rendered. I think it should be just `...plot_quickstart_001.png`?

Author: After changing the path to _images/ as we discussed on the call today, the CircleCI artifacts show all the plots rendering.

If we observe disparities between groups we may want to create a new model
while specifying an appropriate fairness constraint. Note that the choice of
fairness constraints is crucial for the resulting model, and varies based on
application context. If balanced accuracy is highly relevant for fairness in this
contrived example, we can attempt to mitigate the observed disparity using the
corresponding fairness constraint called Equalized Odds. In real world
applications we need to be mindful of the sociotechnical context when making
such decisions. The Exponentiated Gradient mitigation technique used fits the
provided classifier using Equalized Odds as the constraint and Error Rate
as the objective, leading to a vastly reduced difference in  the selection rate:

.. doctest:: quickstart
    :options:  +NORMALIZE_WHITESPACE

    >>> from fairlearn.reductions import ErrorRateParity, EqualizedOdds, ExponentiatedGradient
    >>> np.random.seed(0)  # set seed for consistent results with ExponentiatedGradient

>     >>> objective = ErrorRate(costs={'fp': 0.5, 'fn': 0.5})

Reviewer: I'd love to have someone other than me cross-checking that this makes sense to them.

Author: Miro suggested `objective = ErrorRate(costs={'fp': 0.1, 'fn': 0.9})`, so let me know if that works for you and I can resolve this comment.

    >>> constraint = EqualizedOdds()
    >>> classifier = DecisionTreeClassifier(min_samples_leaf=10, max_depth=4)

>     >>> mitigator = ExponentiatedGradient(classifier, constraint, objective = objective)

Reviewer: ```suggestion
    >>> mitigator = ExponentiatedGradient(classifier, constraint, objective=objective)
```
I know, very nitpicky of me...

    >>> constraint = DemographicParity()
    >>> classifier = DecisionTreeClassifier(min_samples_leaf=10, max_depth=4)

>     >>> mitigator = ExponentiatedGradient(classifier, constraint)

Reviewer: objective="balanced_accuracy_score"

Reviewer: Sorry, that only works for Postprocessing. I don't believe we have a Moment class for balanced accuracy yet, so we @MiroDudik ?

Should be easy to add though... I'll look into it later today.

Reviewer: The `ErrorRate` moment with costs of 0.5 for FP and FN each seems promising. @MiroDudik can you confirm?

Author: Hi @romanlutz and @MiroDudik, any updates on this question?

Author: Can you confirm that ErrorRate is the right moment to use? When I try to use it with Exponentiated Gradient, I am getting `AttributeError: 'ErrorRate' object has no attribute 'default_objective'`

Reviewer: Did you keep the constraint as is and add another arg `objective=ErrorRate(costs={'fp': 0.5, 'fn': 0.5})`? I believe `ErrorRate` wouldn't work as a `constraint` since it doesn't have `default_objective`. 

I suspect you used `ErrorRate` as the `constraint` instead of `DemographicParity`, is that right?

Author: > Did you keep the constraint as is and add another arg `objective=ErrorRate(costs={'fp': 0.5, 'fn': 0.5})`? I believe `ErrorRate` wouldn't work as a `constraint` since it doesn't have `default_objective`.
> 
> I suspect you used `ErrorRate` as the `constraint` instead of `DemographicParity`, is that right?

Gotcha! I pushed the update.

Reviewer: It still uses DP (ie selection rate parity), and not Equalized Odds (FPR/TPR). As I noted in a different comment earlier that is a problem IMO

Author: My apologies, it's been a while since I looked at this file so I walked back too many of the changes. I have changed the constraint back to Equalized Odds.

Reviewer: No worries at all! Ah, the numbers look much more interesting now, too 😄 Before, it was all 0s.
